### PR TITLE
build: modular assembly

### PR DESCRIPTION
**Pull request by itwillwork**
see https://github.com/gravity-ui/rfc/issues/14

### CODE REVIEW

File: .eslintrc [hunk 1]
```diff
@@ -23,6 +23,26 @@
...
+                {
+                    "target": "./src/autocomplete/databases/postgresql/**/*",
+                    "from": "./src/autocomplete/databases/!(postgresql)/**/*"
+                },
+                {
+                    "target": "./src/autocomplete/databases/yql/**/*",
+                    "from": "./src/autocomplete/databases/!(yql)/**/*"
+                },
+            ]
+        }]
```
Comment (itwillwork) [inline .eslintrc:46]: eslint rule for blocking cross imports

File: package.json [hunk 2]
```diff
@@ -9,10 +9,34 @@
...
     "dist/*"
   ],
-  "main": "./dist/index.js",
-  "types": "./dist/index.d.ts",
+  "exports": {
+    "./clickhouse": "./dist/databases/clickhouse/index.js",
+    "./mysql": "./dist/databases/mysql/index.js",
+    "./postgresql": "./dist/databases/postgresql/index.js",
+    "./yql": "./dist/databases/yql/index.js"
+  },
```
Comment (itwillwork) [inline package.json:20]: modular export

File: package.json [hunk 3]
```diff
@@ -33,7 +57,7 @@
     "generate:clickhouse": "rimraf src/autocomplete/databases/clickhouse/generated && antlr4ng -Dlanguage=TypeScript -o src/autocomplete/databases/clickhouse/generated/ -visitor -no-listener -Xexact-output-dir src/autocomplete/databases/clickhouse/grammar/ClickHouseLexer.g4 src/autocomplete/databases/clickhouse/grammar/ClickHouseParser.g4 && bash scripts/patch-generated.sh clickhouse",
     "generate:yql": "rimraf src/autocomplete/databases/yql/generated && antlr4ng -Dlanguage=TypeScript -o src/autocomplete/databases/yql/generated/ -visitor -no-listener -Xexact-output-dir src/autocomplete/databases/yql/grammar/YQL.g4 && bash scripts/patch-generated.sh yql",
     "test": "node --experimental-vm-modules ./node_modules/.bin/jest --verbose src",
-    "build": "rimraf dist && node build.js && tsc -p tsconfig.build.json",
+    "build": "rimraf dist && rimraf dist-types && node build.js && tsc -p tsconfig.build.json --declarationDir dist-types",
```
Comment (roberthovsepyan) [inline package.json:60]: do we need a separate declaration dir? why not just put it in dist?
is this common practice?

Comment (itwillwork) [reply package.json]: no, removed dist-types

File: src/autocomplete/databases/clickhouse/index.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    EngineSuggestion,
+    TableOrViewSuggestion,
+} from '../../shared/autocomplete-types';
+import {clickHouseAutocompleteData} from './clickhouse-autocomplete';
+import {parseQuery, parseQueryWithoutCursor} from '../../shared/autocomplete';
+import {separateQueryAndCursor} from '../../shared/parse-query-with-cursor';
+
+export {ClickHouseLexer} from './generated/ClickHouseLexer';
+
+export {ClickHouseParser} from './generated/ClickHouseParser';
```
Comment (roberthovsepyan) [inline src/autocomplete/databases/clickhouse/index.ts:13]: let's not export this
we don't have a usecase for now

**Review by roberthovsepyan — CHANGES_REQUESTED** [review]

File: build.js [hunk 5]
```diff
@@ -1,11 +1,18 @@
...
-build({
-    entryPoints: ['src/index.ts'],
-    external: ['antlr4ng', 'antlr4-c3'],
-    bundle: true,
-    minify: true,
-    format: 'esm',
-    outfile: 'dist/index.js',
-    tsconfig: './tsconfig.build.json',
-});
+// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
```
[the remark this answers is not in the document]
Comment (itwillwork) [reply build.js:3]: merged

File: src/autocomplete/shared/index.ts [hunk 6]
```diff
@@ -0,0 +1,15 @@
+export * from './autocomplete';
```
[the remark this answers is not in the document]
Comment (itwillwork) [reply src/autocomplete/shared/index.ts:1]: removed

File: src/autocomplete/databases/clickhouse/index.ts [hunk 7]
```diff
@@ -0,0 +1,53 @@
+import {
```
[the remark this answers is not in the document]
Comment (itwillwork) [reply src/autocomplete/databases/clickhouse/index.ts:1]: `(clickhouse | mysql | postgresql | yql)/index.ts` contains public api

for example `mySqlAutocompleteData` from `src/autocomplete/databases/mysql/mysql-autocomplete.ts` usaged in tests `src/autocomplete/shared/cursor.test.ts`.  `mySqlAutocompleteData` should not be in the public api

File: src/autocomplete/databases/clickhouse/index.ts (hunk 4 above)
Comment (itwillwork) [reply src/autocomplete/databases/clickhouse/index.ts:13]: Files are very useful for highlighting semantics errors...Lexer and ...Parser

Comment (Raubzeug) [reply src/autocomplete/databases/clickhouse/index.ts:13]: For error highlight you may use `parsePostgreSqlQueryWithoutCursor(query).errors`

Comment (roberthovsepyan) [reply src/autocomplete/databases/clickhouse/index.ts:13]: better to just use `parseClickHouseQuery`

Comment (itwillwork) [reply src/autocomplete/databases/clickhouse/index.ts]: yes, you're right 

in this case, I can use `parsePostgreSqlQueryWithoutCursor`

deleted

File: build.js [hunk 8]
```diff
@@ -1,12 +1,19 @@
...
-    entryPoints: ['src/index.ts'],
-    external: ['antlr4ng', 'antlr4-c3'],
-    bundle: true,
-    minify: true,
-    keepNames: true,
-    format: 'esm',
-    outfile: 'dist/index.js',
-    tsconfig: './tsconfig.build.json',
-});
+// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
```
Comment (NikitaShkaruba) [inline build.js:3]: Are you sure there's no way to do write it differently? If not, at least write a comment explaining why you need to disable it.

Comment (itwillwork) [reply build.js]: the script has been rewritten, there is no more comment

**Review by roberthovsepyan — APPROVED** [review]

**Pull request merged**